File: estruturasFlexiveis/lista/listaFlexivel.java
package estruturasFlexiveis.lista;

public class listaFlexivel {
    public static void main(String[] args) {

        Lista lista = new Lista();

        // Inserir elementos no início e no fim
        System.out.println("Inserindo elementos...");
        lista.inserirInicio(10);
        lista.inserirFim(20);
        lista.inserirInicio(5);
        lista.inserirFim(30);
        lista.mostrar();

        // Mostrar o tamanho atual da lista
        System.out.println("Tamanho da lista: " + lista.tamanho());  

        // Remover um elemento do início
        System.out.println("Removendo do início: " + lista.removerInicio());  
        System.out.println("Tamanho da lista após remover do início: " + lista.tamanho());
        lista.mostrar();  

        // Remover um elemento do fim
        System.out.println("Removendo do fim: " + lista.removerFim());  
        System.out.println("Tamanho da lista após remover do fim: " + lista.tamanho());
        lista.mostrar();  

        // Inserir em uma posição específica
        System.out.println("Inserindo 15 na posição 1...");
        lista.inserir(15, 1);
        System.out.println("Tamanho da lista após inserir 15: " + lista.tamanho());
        lista.mostrar();    

        // Remover de uma posição específica
        System.out.println("Removendo o elemento da posição 1: " + lista.remover(1)); 
        System.out.println("Tamanho da lista após remover da posição 1: " + lista.tamanho());  
        lista.mostrar();  
    
    }
}

class Celula {
    int elemento;
    Celula prox;

    Celula () {
        this.elemento = 0;
        this.prox = null;
    }

    Celula (int x) {
        this.elemento = x;
        this.prox = null;
    }
}

class Lista {
    Celula primeiro;
    Celula ultimo;

    Lista () {
        primeiro = ultimo = new Celula();
    }

    void inserirFim (int x) {
        ultimo.prox = new Celula(x);
        ultimo = ultimo.prox;
    }

    void inserirInicio (int x) {
        Celula temp = new Celula(x);
        temp.prox = primeiro.prox;
        primeiro.prox = temp;
        if (primeiro.prox == ultimo) {
            ultimo = temp;
        }
        temp = null;
    }

    int removerFim () {
        if (primeiro == ultimo) {
            System.exit(0);
        }
        Celula i;
        for(i=primeiro; i.prox != ultimo; i=i.prox);
        int elemento = ultimo.elemento;
        ultimo = i;
        i = ultimo.prox = null;
        return elemento;
    }

    int removerInicio () { // Remove apenas logicamente e não fisicamente
        if (primeiro == ultimo) {
            System.exit(0);
        }
        Celula temp = primeiro;
        primeiro = primeiro.prox;
        int elemento = primeiro.elemento;
        temp.prox = null;
        temp = null;
        return elemento;
    }

    void inserir (int x, int pos) {
        int tam = tamanho();
        if (pos < 0 || pos > tam) {
            System.exit(0);
        } else if (pos == 0) {
            inserirInicio(x);
        } else if (pos == tam) {
            inserirFim(x);
        } else {
            Celula i = primeiro;
            for(int j=0; j<pos; j++, i=i.prox);
            Celula temp = new Celula(x);
            temp.prox = i.prox;
            i.prox = temp;
            temp = i = null;
        }
    }

    int remover (int pos) {
        int elemento=-1, tam = tamanho();
        if(primeiro == ultimo || pos < 0 || pos >= tam) {
            System.exit(0);
        } else if (pos == 0) {
            elemento = removerInicio();
        } else if (pos == tam-1) {
            elemento = removerFim();
        } else {
            Celula i = primeiro;
            for(int j=0; j<pos; j++, i=i.prox);
            Celula temp = i.prox;
            elemento = temp.elemento;
            i.prox = temp.prox;
            temp.prox = null;
            i = temp = null; 
        }
        return elemento;
    }

    void mostrar () {
        for(Celula i=primeiro.prox; i!=null; i=i.prox) {
            System.out.print(i.elemento + " ");
        }
        System.out.println();
    }

    int tamanho () {
        int tam = 0;
        for(Celula i=primeiro; i!=ultimo; i=i.prox, tam++);
        return tam;
    } 
}

File: estruturas/estruturasFlexiveis/lista/listaFlexivel.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Celula {
    int elemento;
    struct Celula *prox;
    struct Celula *ant;
} Celula ;

typedef struct Lista
{
   struct Celula *primeiro;
   struct Celula *ultimo;
} Lista ;

Lista* inicializar() {
    Lista *lista = (Lista*)malloc(sizeof(Lista));
    lista->primeiro = (Celula*)malloc(sizeof(Celula));
    lista->primeiro->prox = NULL;
    lista->ultimo = lista->primeiro;
    return lista;
}

void inserirInicio (Lista *lista, int x) {
    Celula *temp = (Celula*)malloc(sizeof(Celula));
    temp -> elemento = x;
    temp -> prox = lista -> primeiro -> prox;
    lista -> primeiro -> prox = temp;
    if(lista -> primeiro == lista -> ultimo) {
        lista -> ultimo = temp;
    }
}

void inserirFim(Lista *lista, int x) {
    lista -> ultimo -> prox = (Celula*)malloc(sizeof(Celula));
    lista -> ultimo = lista -> ultimo -> prox;
    lista -> ultimo -> elemento = x;
    lista -> ultimo -> prox = NULL;
}

int removerInicio(Lista *lista) {
    if(lista -> primeiro == lista -> ultimo) {
        printf("LISTA VAZIA");
        exit(1);
    }
    Celula *temp = lista -> primeiro -> prox;
    int elemento = temp ->elemento;
    lista -> primeiro -> prox = temp -> prox;

    if(temp == lista ->ultimo) {
        lista ->ultimo = lista->primeiro;
    }

    temp -> prox = NULL;
    free(temp);
    return elemento;
}

int removerFim(Lista *lista) {
    if(lista -> primeiro == lista -> ultimo) {
        printf("LISTA VAZIA");
        exit(1);
    }
    Celula *i;
    for(i=lista->primeiro; i->prox!=lista->ultimo; i=i->prox);
    int elemento = lista ->ultimo -> elemento;
    free(lista->ultimo);
    lista->ultimo = i;
    lista->ultimo->prox = NULL;
    return elemento;
}

void inserir(Lista *lista, int pos, int x) {
    int tam = tamanho(lista);
    if(pos < 0 || pos > tam) {
        printf("Posicao invalida");
        exit(1);
    }
    if(pos == 0) {
        inserirInicio(lista, x);
    } else if(pos == tam) {
        inserirFim(lista, x);
    } else {
        Celula *j = lista->primeiro;
        for(int i=0; i<pos; i++) {
            j = j->prox;
        }
        Celula *temp = (Celula*)malloc(sizeof(Celula));
        temp -> elemento = x;
        temp -> prox = j->prox;
        j -> prox = temp;
    }
}

int remover (Lista *lista, int pos) {
    int tam = tamanho(lista);
    if(pos < 0 || pos >= tam) {
        printf("Posicao invalida");
        exit(1);
    }
    if(lista -> primeiro == lista -> ultimo) {
        printf("Lista vazia");
        exit(1);
    }
    if(pos == 0) {
        removerInicio(lista);
    } else if(pos == tam-1) {
        removerFim(lista);
    } else {
        Celula *j = lista -> primeiro;
        for(int i=0; i<pos; i++) {
            j = j -> prox;
        }
        Celula *temp = j -> prox;
        int elemento = temp -> elemento;
        j -> prox = temp -> prox;
        temp -> prox = NULL;
        free(temp);
        return elemento;
    }
}

int tamanho(Lista *lista) {
    int cont = 0;
    Celula *i;
    for(i=lista->primeiro->prox; i!=NULL; i=i->prox) {
        cont++;
    }
    return cont;
}

void mostrar(Lista *lista) {
    printf("\n");
    for(Celula *i = lista ->primeiro->prox; i!=NULL; i=i->prox) {
        printf("%d ",i->elemento);
    }
}


int main() {
    Lista *lista = inicializar();
    inserirInicio(lista, 1);
    inserirInicio(lista, 2);
    inserirInicio(lista, 3);
    mostrar(lista);
    inserirFim(lista, 4);
    inserirFim(lista, 5);
    mostrar(lista);
    removerInicio(lista);
    removerFim(lista);
    mostrar(lista);
    inserir(lista, 2, 6);
    remover(lista, 1);
    mostrar(lista);

    return 0;
}